### DISCUSSION TITLE
predict

### DISCUSSION THREAD

**Discussion by AlainZuur** [JenniNiku/gllvm#134] [General]
The second question of the day.

I fit this model:

gl2 <- gllvm(y = SpeciesData, 
             X = CovData,
             row.eff = ~(1 | fDayNum),
             formula = ~ Length.std + Sex,
             num.lv = 1, 
             family = "negative.binomial")

The fitted values are given by:
Fit <- predict(gl2, type = "response", level = 1)   #' Include latent variables.

To verify I fully understand what gllvm is doing, I want to calculate the fitted values myself. For the first species that should be:

Betas <- gl2$params$Xcoef      #betas for covariates
b0     <- gl2$params$beta0       #intercepts
ai      <- gl2$params$row.params   #random effects
theta <- gl2$params$theta             #loadings for 13 species
lv    <- gl2$lvs  #' getLV(gl2, type = "residual")    #latent variable..only 1 in this model
X     <- model.matrix(~ Length.std + Sex, data = Fish2)    #X matrix for covariates
Day   <- as.numeric(CovData$fDayNum)                      #Index for random effects

And here I calculate the fitted values manually (for the first species):

eta   <- ai[Day] + X %*% c(b0[1], Betas[1,]) + lv * theta[1]
MyFit <- exp(eta)

Why are Fit[,1] and MyFit not the same? If I use  num.lv = 0, then they are. I think the question comes down to: what exactly is  predict(gl2, type = "response", level = 1) doing?

If it helps to send data and condensed R code, give a shout.
Kind regards,

Alain

**Comment by BertvanderVeen** [comment]
Thanks Alain.

For technical reasons (compatibility with generalized reduced rank regression/constrained ordination and concurrent ordination) the main diagonal elements of the loadings are fixed to one. These are separately stored in gl2$params$sigma.lv, and need to be column-wise (i.e., first entry of sigma.lv multiplies with first column of the loadings etc.) multiplied with the loadings. Have a look and see if that fixes your issue.

**Reply by AlainZuur** [reply]
Excellent.
theta <- gl2$params$theta  * gl2$params$sigma.lv   #' In case there is one latent variable.

Now it all matches up.

Alain
